core/src/main/java/org/apache/iceberg/CherryPickOperation.java

```diff
@@ -0,0 +1,231 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.List;
+import java.util.Map;
+import org.apache.iceberg.exceptions.CherrypickAncestorCommitException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.PartitionSet;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.apache.iceberg.util.WapUtil;
+
+/**
+ * Cherry-picks or fast-forwards the current state to a snapshot.
+ * <p>
+ * This update is not exposed though the Table API. Instead, it is a package-private part of the Transaction API
+ * intended for use in {@link ManageSnapshots}.
+ */
+class CherryPickOperation extends MergingSnapshotProducer<CherryPickOperation> {
+
+  private final Map<Integer, PartitionSpec> specsById;
+  private Snapshot cherrypickSnapshot = null;
+  private boolean requireFastForward = false;
+  private PartitionSet replacedPartitions = null;
+
+  CherryPickOperation(String tableName, TableOperations ops) {
+    super(tableName, ops);
+    this.specsById = ops.current().specsById();
+  }
+
+  @Override
+  protected CherryPickOperation self() {
+    return this;
+  }
+
+  @Override
+  protected String operation() {
+    // snapshotOperation is used by SnapshotProducer when building and writing a new snapshot for cherrypick
+    Preconditions.checkNotNull(cherrypickSnapshot, "[BUG] Detected uninitialized operation");
+    return cherrypickSnapshot.operation();
+  }
+
+  public CherryPickOperation cherrypick(long snapshotId) {
+    TableMetadata current = current();
+    this.cherrypickSnapshot = current.snapshot(snapshotId);
+    ValidationException.check(cherrypickSnapshot != null,
+        "Cannot cherry-pick unknown snapshot ID: %s", snapshotId);
+
+    if (cherrypickSnapshot.operation().equals(DataOperations.APPEND)) {
+      // this property is set on target snapshot that will get published
+      String wapId = WapUtil.validateWapPublish(current, snapshotId);
+      if (wapId != null) {
+        set(SnapshotSummary.PUBLISHED_WAP_ID_PROP, wapId);
+      }
+
+      // link the snapshot about to be published on commit with the picked snapshot
+      set(SnapshotSummary.SOURCE_SNAPSHOT_ID_PROP, String.valueOf(snapshotId));
+
+      // Pick modifications from the snapshot
+      for (DataFile addedFile : cherrypickSnapshot.addedFiles()) {
+        add(addedFile);
+      }
+
+    } else if (cherrypickSnapshot.operation().equals(DataOperations.OVERWRITE) &&
+        PropertyUtil.propertyAsBoolean(cherrypickSnapshot.summary(), SnapshotSummary.REPLACE_PARTITIONS_PROP, false)) {
+      // the operation was ReplacePartitions. this can be cherry-picked iff the partitions have not been modified.
```

Comment:
typo: s/iff/if

Reply:
This is actually math shorthand for [if and only if](https://en.wikipedia.org/wiki/If_and_only_if), not a typo.

Reply:
Ah I see, doh!